signer/xpi/jar.go

```diff
@@ -0,0 +1,173 @@
+package xpi
+
+import (
+	"archive/zip"
+	"bytes"
+	"crypto/sha1"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strings"
+)
+
+func makeJARManifests(input []byte) (manifest, sigfile []byte, err error) {
+	inputReader := bytes.NewReader(input)
+	r, err := zip.NewReader(inputReader, int64(len(input)))
+	if err != nil {
+		return
+	}
+
+	// first generate the manifest file by calculated a sha256 in each zip entry
+	mw := bytes.NewBuffer(manifest)
+
+	for _, f := range r.File {
+		if isSignatureFile(f.Name) {
+			// reserved signature files do not get included in the manifest
+			continue
+		}
+		if f.FileInfo().IsDir() {
+			// directories do not get included
+			continue
+		}
+		rc, err := f.Open()
+		if err != nil {
+			return manifest, sigfile, err
+		}
+		data, err := ioutil.ReadAll(rc)
+		if err != nil {
+			return manifest, sigfile, err
+		}
+		fmt.Fprintf(mw, "Name: %s\nDigest-Algorithms: SHA1 SHA256\n", f.Name)
+		h1 := sha1.New()
+		h1.Write(data)
+		fmt.Fprintf(mw, "SHA1-Digest: %s\n", base64.StdEncoding.EncodeToString(h1.Sum(nil)))
+		h2 := sha256.New()
+		h2.Write(data)
+		fmt.Fprintf(mw, "SHA256-Digest: %s\n\n", base64.StdEncoding.EncodeToString(h2.Sum(nil)))
+	}
+	manifestBody := mw.Bytes()
+	manifest = []byte(`Manifest-Version: 1.0
+
+`)
```

Comment:
It isn't necessarily clear if that newline is important or just accidental and should be commented or otherwise added (e.g escaped and via `\n` or so)